server/render.js

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const Eleventy = require('@11ty/eleventy');
+const path = require('path');
+const yaml = require('js-yaml');
+const xss = require('xss').filterXSS;
+const {EleventyRenderPlugin} = require('@11ty/eleventy');
+const {OAuth2Client} = require('google-auth-library');
+
+const ignorePatterns = [
+  'site/es/**/*',
+  'site/pt/**/*',
+  'site/ko/**/*',
+  'site/zh/**/*',
+  'site/ru/**/*',
+  'site/ja/**/*',
+  'site/fr/**/*',
+  'site/de/**/*',
+  'site/**/docs/',
+  'site/**/articles/',
+  'site/**/authors/',
+  'site/**/tags/',
+  'site/**/releases/',
+  'site/**/workbox/',
+];
+
+const CLIENT_ID =
+  '636061184119-c4pof3b538r4ibs8bt6g8qdfth958nc9.apps.googleusercontent.com';
+const authClient = new OAuth2Client(CLIENT_ID);
+
+async function verifyLogin(token) {
+  const ticket = await authClient.verifyIdToken({
+    idToken: token,
+    audience: CLIENT_ID,
+  });
+  const payload = ticket.getPayload();
+  return payload && payload['sub']; // userid
+}
+
+/**
+ * Escapes and formats frontmatter provided by user to 11ty format.
+ * @param {string} inputFm Frontmatter string
+ */
+const parseFrontmatter = inputFm => {
+  const /** @type {any} */ parsedFm = yaml.load(inputFm) || {};
+  const date = parsedFm.date?.toISOString ? parsedFm.date : new Date();
+  return {
+    layout: xss(parsedFm.layout) || 'layouts/blog-post.njk',
+    date: xss(date.toISOString().split('T')[0]),
+    title: xss(parsedFm.title) || 'Test title TBD',
+    desciption: xss(parsedFm.description) || 'Test description TBD',
+    authors: parsedFm.authors?.map(author => xss(author)) || ['rachelandrew'],
+  };
+};
+
+const supportedLayouts = ['layouts/blog-post.njk', 'layouts/doc-post.njk'];
+
+const renderHandler = async (req, res) => {
+  // Check sign in
+  const userId = await verifyLogin(req.body.t).catch(() => {
+    res.sendStatus(403);
+  });
+  if (!userId) {
+    return;
+  }
+
+  process.env.ELEVENTY_IGNORE_EXTENSIONS = 'true';
+
+  let fm = {};
+  try {
+    fm = parseFrontmatter(req.body.fm);
+  } catch (e) {
+    return res.status(400).send(`FrontMatter malformed: ${e}`);
+  }
+  if (!supportedLayouts.includes(fm.layout)) {
+    return res.status(400).send(`Unsupported layout: ${fm.layout}`);
+  }
+  const body = xss(req.body.md);
+  const inputPath = path.join('./site/en/preview', fm.layout);
+
+  try {
+    const elev = new Eleventy(inputPath, 'dist', {
+      configPath: '.eleventy.js',
+      inputDir: './site',
+      config: function (eleventyConfig) {
+        eleventyConfig.addPlugin(EleventyRenderPlugin);
+        eleventyConfig.addGlobalData('fm', fm);
+        eleventyConfig.addGlobalData('body', body);
+        eleventyConfig.addGlobalData('date', fm.date);
```

Comment:
Not sure what is cleaner/faster/easier to read but you could probably also manually write them to `this.ctx` in the render shortcode...?

Reply:
Actually it does not work when set directly on this.ctx object. I don't know why, but I'd say passing it as global data is consistent enough with the rest of the site.